tests/Internal/Codebase/ClassConstantByWildcardResolverTest.php

```diff
@@ -0,0 +1,83 @@
+<?php
+declare(strict_types=1);
+
+namespace Psalm\Tests\Internal\Codebase;
+
+use Psalm\Internal\Codebase\ClassConstantByWildcardResolver;
+use Psalm\Tests\TestCase;
+use Psalm\Type\Atomic\TLiteralString;
+
+final class ClassConstantByWildcardResolverTest extends TestCase
+{
+    /**
+     * @var \Psalm\Internal\Codebase\ClassConstantByWildcardResolver
+     */
+    private $resolver;
+
+    public function setUp(): void
+    {
+        parent::setUp();
+        $this->resolver = new ClassConstantByWildcardResolver($this->project_analyzer->getCodebase());
+    }
+
+    public function testWillParseAllClassConstants(): void
+    {
+        $this->addFile(
+            'psalm-assert.php',
+            '
+            <?php
+            namespace ReconciliationTest;
+            class Foo
+            {
+                const PREFIX_BAR = \'bar\';
+                const PREFIX_BAZ = \'baz\';
+                const PREFIX_QOO = Foo::PREFIX_BAR;
+            }
+            '
+        );
+        $this->project_analyzer->getCodebase()->scanFiles();
+        $resolved = $this->resolver->resolve('ReconciliationTest\\Foo', '*');
+        self::assertNotEmpty($resolved);
+        foreach ($resolved as $type) {
+            self::assertInstanceOf(TLiteralString::class, $type);
+            self::assertTrue($type->value === 'bar' || $type->value === 'baz');
+        }
+    }
+
+    public function testWillParseMatchingClassConstants(): void
+    {
+        $this->addFile(
+            'psalm-assert.php',
+            '
+            <?php
+            namespace ReconciliationTest;
+            class Foo
+            {
+                const BAR = \'bar\';
+                const BAZ = \'baz\';
+                const QOO = \'qoo\';
+            }
+            '
+        );
+        $this->project_analyzer->getCodebase()->scanFiles();
+        $resolved = $this->resolver->resolve('ReconciliationTest\\Foo', 'BA*');
+        self::assertNotEmpty($resolved);
+        foreach ($resolved as $type) {
+            self::assertInstanceOf(TLiteralString::class, $type);
+            self::assertTrue($type->value === 'bar' || $type->value === 'baz');
+        }
+
+        $resolved = $this->resolver->resolve('ReconciliationTest\\Foo', 'QOO');
+        self::assertNotNull($resolved);
+        /** @var list<\Psalm\Type\Atomic> */
+        self::assertCount(1, $resolved);
+        /** @var non-empty-list<\Psalm\Type\Atomic> $type */
```

Comment:
Why do you set $type as an array here, just to assert that it's an instance of TLiteralString just below?

Reply:
because the "shepherd" build pipeline step yelled at me that accessing an undefined array key would be problematic...

Reply:
Here are ways you should be able to solve this correctly:
```
/** @var non-empty-list<\Psalm\Type\Atomic> $resolved */
$type = $resolved[0];
self::assertInstanceOf(TLiteralString::class, $type);
```
or
```
/** @var \Psalm\Type\Atomic $type */
$type = $resolved[0];
self::assertInstanceOf(TLiteralString::class, $type);
```

in any case, $type can't be an array AND then a class-string, and you shouldn't need to suppress any DocblockTypeContradiction

Reply:
Yah nvm, dunno why - I've removed the whole annotations and now it works.
I guess there was some errors due to the `var` annotations.